src/psyclone/domain/lfric/lfric_extract_driver_creator.py

```diff
@@ -0,0 +1,974 @@
+
+        # First handle variables that are read:
+        # -------------------------------------
+        for signature in input_list:
+            # Find the right symbol for the variable. Note that all variables
+            # in the input and output list have been detected as being used
+            # when the variable accesses were analysed. Therefore, these
+            # variables have References, and will already have been declared
+            # in the symbol table (in add_all_kernel_symbols).
+            sig_str = LFRicExtractDriverCreator.flatten_string(str(signature))
+            orig_sym = original_symbol_table.lookup(signature[0])
+            if orig_sym.is_array and orig_sym.datatype.intrinsic.name in \
+                    self._all_field_types:
+                upper = int(orig_sym.datatype.shape[0].upper.value)
+                for i in range(1, upper+1):
+                    sym = symbol_table.lookup_with_tag(f"{sig_str}%{i}")
+                    name_lit = Literal(f"{sig_str}%{i}", CHARACTER_TYPE)
+                    LFRicExtractDriverCreator.add_call(program, read_var,
+                                                       [name_lit,
+                                                        Reference(sym)])
+                continue
+
+            sym = symbol_table.lookup_with_tag(str(signature))
+            name_lit = Literal(str(signature), CHARACTER_TYPE)
+            LFRicExtractDriverCreator.add_call(program, read_var,
+                                               [name_lit, Reference(sym)])
+
+        # Then handle all variables that are written (note that some
+        # variables might be read and written)
+        # ----------------------------------------------------------
+        # Collect all output symbols to later create the tests for
+        # correctness. This list stores 2-tuples: first one the
+        # variable that stores the output from the kernel, the second
+        # one the variable that stores the output values read from the
+        # file. The content of these two variables should be identical
+        # at the end.
+        output_symbols = []
+        for signature in output_list:
+            # Find the right symbol for the variable. Note that all variables
+            # in the input and output list have been detected as being used
+            # when the variable accesses were analysed. Therefore, these
+            # variables have References, and will already have been declared
+            # in the symbol table (in add_all_kernel_symbols).
+            orig_sym = original_symbol_table.lookup(signature[0])
+            is_input = signature in input_list
+            if orig_sym.is_array and orig_sym.datatype.intrinsic.name in \
+                    self._all_field_types:
+                flattened = LFRicExtractDriverCreator.\
+                    flatten_string(str(signature))
+                upper = int(orig_sym.datatype.shape[0].upper.value)
```

Comment:
Yes, and again I think this must be the case, since PSyclone would otherwise not know how many arguments to add to a kernel call for an array of fields.